### PR TITLE
Dockerfile refactored to use Alpine.

### LINKED ISSUE #3782: Rewrite our Dockerfile
Our current Dockerfile was largely written roughly 6 months before Let's Encrypt entered beta and it was a convenient way to test the project on different systems. It's in need of an update. I propose we update our Dockerfile to do the following:

* Continue installing from our repo rather than OS packages
* Remove unneeded OS dependencies after installing Certbot (such as `gcc`?)
* Clean `pip` cache after installing `certbot` (or don't use it all)
* Don't install the Apache or Nginx plugins

The resulting image should be roughly the size if we simply installed the `certbot` OS package. If this isn't possible for some reason, we may have to choose a different approach.

### PR DESCRIPTION
**Pull request by Yajo**
Fix #3782, and hopefully help #492. Idea from #3929.

@Tecnativa

### CODE REVIEW

**Comment by willfarrell** [comment]
Leveraging what will be in the certbot-auto, you could simplify it even more.

```Dockerfile
FROM alpine:3.5

RUN apk --no-cache add \
        bash \
        openssl \
        wget

# certbot
RUN cd /usr/bin \
    && wget https://dl.eff.org/certbot-auto \
    && chmod a+x certbot-auto \
    && certbot-auto --version --non-interactive
...
```

**Comment by Yajo** [comment]
Cool idea, although I did not want to make this PR dependant on #3929. After that one is merged, this one could change, but this way it is not a requirement.

**Comment by hamburml** [comment]
Any news? Would like to have a smaller container :)

**Comment by willfarrell** [comment]
I actually switched to using dehydrated, smaller and has dns-01 and ecdsa support build it. You can find the docker image here: https://hub.docker.com/r/willfarrell/letsencrypt/

**Comment by hamburml** [comment]
@willfarrell Thanks. Would also love to use an official image with tags pushed to docker hub for business usage.

File: Dockerfile [hunk 1]
```diff
@@ -1,70 +1,30 @@
...
-# install in editable mode (-e) to save space: it's not possible to
-# "rm -rf /opt/certbot/src" (it's stays in the underlaying image);
-# this might also help in debugging: you can "docker run --entrypoint
-# bash" and investigate, apply patches, etc.
-
 ENTRYPOINT [ "certbot" ]
+COPY . src
+
+RUN apk add --no-cache --virtual .certbot-deps \
+        dialog \
```
Comment (bmw) [inline Dockerfile]: This dependency was removed in Certbot 0.10.0. Looks like we forgot to update our current Dockerfile.

File: Dockerfile [hunk 2]
```diff
@@ -1,70 +1,30 @@
...
+        augeas-libs \
```
Comment (bmw) [inline Dockerfile]: If `certbot-apache` is no longer installed, this package can also be removed.

File: Dockerfile [hunk 3]
```diff
@@ -1,70 +1,30 @@
...
+        libffi \
+        libssl1.0 \
+        wget \
```
Comment (bmw) [inline Dockerfile]: Why is this package needed?

**Comment by Yajo** [comment]
Thanks for your comments @bmw, I agree on all of them. I just wanted to provide something similar to what you had before, but more Docker-ish and Alpine-based; indeed, I think it's best that you, with more certbot knowledge than me, take this base and build on it what you want.

**Comment by Yajo** [comment]
Cool! It would be great to trigger an automated build in the docker hub and link it in the readme, don't you think?

**Pull request merged**